Fix issue #7205: [Feature]: Allow repo microagent Markdown Files Without Required Header

openhands-agent opened the pull request:
This pull request fixes #7205.

The issue has been successfully resolved based on the changes made. The key modifications directly address the core problem:

1. The BaseMicroAgent.load() method was modified to handle files without frontmatter by implementing default values, specifically:
- Setting type to MicroAgentType.REPO_KNOWLEDGE when no type is specified
- Using 'repo' as the default name for repository agents
- Maintaining other required metadata fields with sensible defaults

2. The implementation allows for multiple scenarios:
- Files with no frontmatter at all
- Files with empty frontmatter
- Files with partial frontmatter
- Files with complete frontmatter (maintaining backward compatibility)

3. Comprehensive test coverage was added that verifies all these scenarios work as expected, with tests specifically checking that markdown files load correctly with and without headers.

4. Documentation was updated to reflect the new flexibility, clearly stating that frontmatter is optional for repository agents while maintaining existing functionality for other agent types.

These changes directly solve the original issue by removing the requirement for headers in Markdown files while ensuring the system continues to function properly, meeting the user experience goals outlined in the issue description. The solution is complete and self-contained, requiring no additional modifications to work as intended.

Automatic fix generated by [OpenHands](https://github.com/All-Hands-AI/OpenHands/) 🙌

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:df63204-nikolaik   --name openhands-app-df63204   docker.all-hands.dev/all-hands-ai/openhands:df63204
```

openhands-agent commented:
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/14222042644).

Prateikx commented on openhands/microagent/microagent.py:
```diff
@@ -46,8 +46,24 @@ def load(
         loaded = frontmatter.load(file_io)
         content = loaded.content
+
+        # Handle case where there's no frontmatter or empty frontmatter
```
**✨ AI Suggestion:** The `metadata_dict` is unnecessarily initialized with an empty dictionary and then populated. It's more efficient to directly use `loaded.metadata` with a default value if it's empty.
```suggestion
+        # Handle case where there's no frontmatter or empty frontmatter
+        metadata_dict = loaded.metadata or {}
+
+        # If no type is specified, default to repo type
+        if 'type' not in metadata_dict:
+            metadata_dict['type'] = MicroAgentType.REPO_KNOWLEDGE
+
+        # If no name is specified, use 'repo' for repo type
+        if 'name' not in metadata_dict:
+            metadata_dict['name'] = 'repo' if metadata_dict['type'] == MicroAgentType.REPO_KNOWLEDGE else 'default'
+
```

enyst replied:
Thanks, fixed!

enyst commented on openhands/microagent/microagent.py:
```diff
@@ -46,8 +46,24 @@ def load(
         content = loaded.content
+
+        # Handle case where there's no frontmatter or empty frontmatter
+        metadata_dict = loaded.metadata if loaded.metadata else {}
```
```suggestion
        metadata_dict = loaded.metadata or {}
```

enyst commented:
@openhands-agent Read the description and the diff of this PR, PR 7655 and understand its goal. Use the GitHub API to read, you have a token in your env.

Then look at MicroAgentMetadata class and see its fields, and their default values. Set the default value for `type` to REPO_KNOWLEDGE, and that will allow you to remove useless code from the PR diff, from the microagent.py file which is effectively trying to set a default value for `type`.

Look at the diff carefully and see if there is anything else you can simplify. Then do it.

You don't need to run tests right now, I took care of it in CI.

openhands-agent commented:
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/14229353799).

openhands-agent commented:
Here's a concise overview of the changes:

CHANGES MADE:
- Removed requirement for headers in Markdown files
- Moved default type settings to MicroAgentMetadata
- Eliminated redundant code for handling missing headers
- Implemented robust default values

STATUS:
✅ All key issues appear to be successfully resolved:
- Headers are now optional
- Default handling is centralized
- Backward compatibility maintained
- Code complexity reduced

OUTCOME:
The changes effectively simplified the codebase while improving user experience through optional headers and robust defaults. No remaining issues are apparent from the provided information.

enyst commented on tests/unit/test_microagent_no_header.py:
```diff
@@ -0,0 +1,94 @@
+
+    # Verify it's loaded as a repo agent with default values
+    assert isinstance(agent, RepoMicroAgent)
+    assert agent.name == 'repo'
```
```suggestion
    assert agent.name == 'default'
```

enyst commented on tests/unit/test_microagent_no_header.py:
```diff
@@ -0,0 +1,94 @@
+
+    # Verify it's loaded as a repo agent with default values
+    assert isinstance(agent, RepoMicroAgent)
+    assert agent.name == 'repo'
```
```suggestion
    assert agent.name == 'default'
```
